src/MongoDB.Driver/Linq/Linq3Implementation/Ast/PipelineOptimizer/AstGroupPipelineOptimizer.cs

```diff
@@ -0,0 +1,413 @@
+                {
+                    var stage = pipeline.Stages[j];
+                    if (StageCanBeOptimized(stage))
+                    {
+                        stages.Add(stage);
+                    }
+
+                    if (IsLastStageThatCanBeOptimized(stage))
+                    {
+                        return stages;
+                    }
+                }
+
+                return null;
+
+                static bool StageCanBeOptimized(AstStage stage)
+                {
+                    return stage.NodeType switch
+                    {
+                        AstNodeType.LimitStage => true,
+                        AstNodeType.MatchStage => true,
+                        AstNodeType.ProjectStage => true,
+                        AstNodeType.SampleStage => true,
+                        AstNodeType.SkipStage => true,
+                        _ => false
+                    };
+                }
+
+                static bool IsLastStageThatCanBeOptimized(AstStage stage)
+                {
+                    return stage.NodeType switch
+                    {
+                        AstNodeType.ProjectStage => true,
+                        _ => false
+                    };
+                }
+            }
+        }
+
+        private (AstNode, AstNode)[] OptimizeGroupAndFollowingStages(AstGroupStage groupStage, List<AstStage> followingStages)
+        {
+            var mappings = new List<(AstNode, AstNode)>();
+
+            foreach (var stage in followingStages)
+            {
+                var optimizedStage = OptimizeFollowingStage(stage);
+                if (optimizedStage != stage)
+                {
+                    mappings.Add((stage, optimizedStage));
+                }
+            }
+
+            if (_accumulators.Count > 0)
+            {
+                var newGroupStage = AstStage.Group(groupStage.Id, _accumulators);
+                mappings.Add((groupStage, newGroupStage));
+            }
+
+            return mappings.ToArray();
+        }
+
+        private AstStage OptimizeFollowingStage(AstStage stage)
+        {
+            return stage switch
+            {
+                AstLimitStage limitStage => OptimizeLimitStage(limitStage),
+                AstMatchStage matchStage => OptimizeMatchStage(matchStage),
+                AstProjectStage projectStage => OptimizeProjectStage(projectStage),
+                AstSampleStage sampleStage => OptimizeSampleStage(sampleStage),
+                AstSkipStage skipStage => OptimizeSkipStage(skipStage),
+                _ => throw new InvalidOperationException($"Unexpected node type: {stage.NodeType}.")
+            };
+        }
+
+        private AstStage OptimizeLimitStage(AstLimitStage stage)
+        {
+            return stage;
+        }
+
+        private AstStage OptimizeMatchStage(AstMatchStage stage)
+        {
+            var optimizedFilter = AccumulatorMover.MoveAccumulators(_accumulators, stage.Filter);
```

Comment:
The `AccumulatorMover` didn't really need a reference to `this`. It only needs a reference to the set of accumulators.